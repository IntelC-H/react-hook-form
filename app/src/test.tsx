import * as React from 'react';
import { useFieldArray, useForm } from 'react-hook-form';
import { Control } from '../../src/types';

type FormData = {
  nest: {
    test: {
      value: string;
      nestedArray: {
        value: string;
      }[];
    }[];
  };
};
const ChildComponent = ({
  index,
  control,
}: {
  control: Control<FormData>;
  index: number;
}) => {
  const { fields } = useFieldArray<FormData>({
    name: `nest.test.${index}.nestedArray` as const,
    control,
  });

  return (
    <div>
      {fields.map((item, i) => (
        <input
          key={item.id}
          {...control.register(
            `nest.test.${index}.nestedArray.${i}.value` as const,
          )}
        />
      ))}
    </div>
  );
};

const Component = () => {
  const { register, control } = useForm({
    defaultValues: {
      nest: {
        test: [
          { value: '1', nestedArray: [{ value: '2' }] },
          { value: '3', nestedArray: [{ value: '4' }] },
        ],
      },
    },
  });
  const { fields, remove, append } = useFieldArray({
    name: 'nest.test',
    control,
  });

  return (
    <div>
      {fields.map((item, i) => (
        <div key={item.id}>
          <input {...register(`nest.test.${i}.value` as const)} />

          <ChildComponent control={control} index={i} />

          <button
            type={'button'}
            onClick={() => remove(i)}
            data-testid={item.value}
          >
            remove
          </button>
        </div>
      ))}

      <button type={'button'} onClick={() => append({ value: 'test' })}>
        append
      </button>
    </div>
  );
};

export default Component;
