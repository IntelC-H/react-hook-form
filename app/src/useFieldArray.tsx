import React from 'react';
import {
  useForm,
  useFieldArray,
  Controller,
  ErrorMessage,
} from 'react-hook-form';

let renderCount = 0;

const UseFieldArray: React.FC = (props: any) => {
  const {
    control,
    handleSubmit,
    register,
    formState: { dirty, touched, isValid, dirtyFields },
    reset,
    errors,
  } = useForm<{
    data: { name: string }[];
  }>({
    ...(props.match.params.mode === 'default'
      ? {
          defaultValues: {
            data: [{ name: 'test' }, { name: 'test1' }, { name: 'test2' }],
          },
        }
      : {}),
    mode: props.match.params.mode === 'formState' ? 'onChange' : 'onSubmit',
  });
  const {
    fields,
    append,
    prepend,
    swap,
    move,
    insert,
    remove,
  } = useFieldArray<{ name: string }>({
    control,
    name: 'data',
  });
  const [data, setData] = React.useState([]);
  const onSubmit = (data: any) => {
    setData(data);
  };

  React.useEffect(() => {
    setTimeout(() => {
      if (props.match.params.mode === 'asyncReset') {
        reset({
          data: [{ name: 'test' }, { name: 'test1' }, { name: 'test2' }],
        });
      }
    }, 10);
  }, [reset, props.match.params.mode]);

  renderCount++;

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <ul>
        {fields.map((data, index) => (
          <li key={data.id}>
            {index % 2 ? (
              <input
                id={`field${index}`}
                name={`data[${index}].name`}
                defaultValue={data.name}
                data-order={index}
                ref={register({ required: 'This is required' })}
              />
            ) : (
              <Controller
                as={<input id={`field${index}`} />}
                control={control}
                rules={{
                  required: 'This is required',
                }}
                name={`data[${index}].name`}
                defaultValue={data.name}
                data-order={index}
              />
            )}
            <ErrorMessage errors={errors} name={`data[${index}].name`}>
              {({ message }) => <p id={`error${index}`}>{message}</p>}
            </ErrorMessage>
            <button id={`delete${index}`} onClick={() => remove(index)}>
              Delete
            </button>
          </li>
        ))}
      </ul>

      <button
        id="append"
        type="button"
        onClick={() => append({ name: renderCount.toString() })}
      >
        append
      </button>

      <button
        id="prepend"
        type="button"
        onClick={() => prepend({ name: renderCount.toString() })}
      >
        prepend
      </button>

      <button id="swap" onClick={() => swap(1, 2)} type="button">
        swap
      </button>

      <button id="move" onClick={() => move(2, 0)} type="button">
        move
      </button>

      <button
        id="insert"
        type="button"
        onClick={() => insert(1, { name: renderCount.toString() })}
      >
        insert
      </button>

      <button id="remove" type="button" onClick={() => remove(1)}>
        remove
      </button>

      <button id="removeAll" type="button" onClick={() => remove()}>
        remove all
      </button>

      <button id="submit">Submit</button>

      <div id="renderCount">{renderCount}</div>
      <div id="result">{JSON.stringify(data)}</div>
      <div id="dirty">{dirty ? 'yes' : 'no'}</div>
      <div id="isValid">{isValid ? 'yes' : 'no'}</div>
      <div id="dirtyFields">{JSON.stringify([...dirtyFields])}</div>
      <div id="touched">{JSON.stringify(touched.data)}</div>
    </form>
  );
};

export default UseFieldArray;
