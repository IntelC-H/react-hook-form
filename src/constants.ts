import { ValidationMode } from './types';

export const VALIDATION_MODE: ValidationMode = {
  onBlur: 'onBlur',
  onChange: 'onChange',
  onSubmit: 'onSubmit',
};

export const RADIO_INPUT = 'radio';
