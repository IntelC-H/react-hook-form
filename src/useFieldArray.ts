import * as React from 'react';
import { useFormContext } from './useFormContext';
import { isMatchFieldArrayName } from './logic/isNameInFieldArray';
import getFieldValueByName from './logic/getFieldValueByName';
import { appendId, mapIds } from './logic/mapIds';
import getIsFieldsDifferent from './logic/getIsFieldsDifferent';
import get from './utils/get';
import isUndefined from './utils/isUndefined';
import removeArrayAt from './utils/remove';
import moveArrayAt from './utils/move';
import swapArrayAt from './utils/swap';
import prependAt from './utils/prepend';
import isArray from './utils/isArray';
import insertAt from './utils/insert';
import fillEmptyArray from './utils/fillEmptyArray';
import {
  Field,
  FieldValues,
  UseFieldArrayProps,
  Control,
  ArrayField,
} from './types';

const { useEffect, useCallback, useRef, useState } = React;

export const useFieldArray = <
  FormArrayValues extends FieldValues = FieldValues,
  KeyName extends string = 'id',
  ControlProp extends Control = Control
>({
  control,
  name,
  keyName = 'id' as KeyName,
}: UseFieldArrayProps<KeyName, ControlProp>) => {
  const methods = useFormContext();
  const {
    resetFieldArrayFunctionRef,
    fieldArrayNamesRef,
    reRender,
    fieldsRef,
    getValues,
    defaultValuesRef,
    removeFieldEventListener,
    errorsRef,
    dirtyFieldsRef,
    isDirtyRef,
    touchedFieldsRef,
    readFormStateRef,
    watchFieldsRef,
    validFieldsRef,
    fieldsWithValidationRef,
    validateSchemaIsValid,
  } = control || methods.control;
  const memoizedDefaultValues = useRef(get(defaultValuesRef.current, name, []));
  const [fields, setField] = useState<
    Partial<ArrayField<FormArrayValues, KeyName>>[]
  >(mapIds(memoizedDefaultValues.current, keyName));
  const allFields = useRef(fields);

  allFields.current = fields;

  const appendValueWithKey = (value: Partial<FormArrayValues>[]) =>
    value.map((v: Partial<FormArrayValues>) => appendId(v, keyName));

  const commonTasks = (fieldsValues: any) => {
    setField(fieldsValues);

    if (readFormStateRef.current.isValid && validateSchemaIsValid) {
      validateSchemaIsValid({
        [name]: fieldsValues,
      });
    }
  };

  const resetFields = (flagOrFields?: Partial<FormArrayValues>[]) => {
    if (readFormStateRef.current.dirty) {
      isDirtyRef.current = isUndefined(flagOrFields)
        ? true
        : getIsFieldsDifferent(flagOrFields, memoizedDefaultValues.current);
    }

    for (const key in fieldsRef.current) {
      if (isMatchFieldArrayName(key, name) && fieldsRef.current[key]) {
        removeFieldEventListener(fieldsRef.current[key] as Field, true);
      }
    }
  };

  const mapCurrentFieldsValueWithState = () => {
    const currentFieldsValue = getValues({ nest: true })[name];

    if (isArray(currentFieldsValue)) {
      for (let i = 0; i < currentFieldsValue.length; i++) {
        allFields.current[i] = {
          ...allFields.current[i],
          ...currentFieldsValue[i],
        };
      }
    }
  };

  const append = (
    value: Partial<FormArrayValues> | Partial<FormArrayValues>[],
  ) => {
    if (readFormStateRef.current.dirty) {
      isDirtyRef.current = true;
      reRender();
    }
    commonTasks([
      ...allFields.current,
      ...(isArray(value)
        ? appendValueWithKey(value)
        : [appendId(value, keyName)]),
    ]);
  };

  const prepend = (
    value: Partial<FormArrayValues> | Partial<FormArrayValues>[],
  ) => {
    resetFields();
    commonTasks(
      prependAt(
        allFields.current,
        isArray(value) ? appendValueWithKey(value) : [appendId(value, keyName)],
      ),
    );

    if (errorsRef.current[name]) {
      errorsRef.current[name] = prependAt(
        errorsRef.current[name],
        fillEmptyArray(value),
      );
    }

    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      touchedFieldsRef.current[name] = prependAt(
        touchedFieldsRef.current[name],
        fillEmptyArray(value),
      );
      reRender();
    }
  };

  const remove = (index?: number | number[]) => {
    let shouldRender = false;
    if (!isUndefined(index)) {
      mapCurrentFieldsValueWithState();
    }

    resetFields(
      removeArrayAt(getFieldValueByName(fieldsRef.current, name), index),
    );
    commonTasks(removeArrayAt(allFields.current, index));

    if (errorsRef.current[name]) {
      errorsRef.current[name] = removeArrayAt(errorsRef.current[name], index);
      if (!(errorsRef.current[name] as []).filter(Boolean).length) {
        delete errorsRef.current[name];
      }
    }

    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      touchedFieldsRef.current[name] = removeArrayAt(
        touchedFieldsRef.current[name],
        index,
      );
      shouldRender = true;
    }

    if (readFormStateRef.current.dirty) {
      dirtyFieldsRef.current.forEach(dirtyField => {
        if (isUndefined(name) || dirtyField.startsWith(`${name}[${index}]`)) {
          dirtyFieldsRef.current.delete(dirtyField);
        }
      });
      shouldRender = true;
    }

    if (readFormStateRef.current.isValid && !validateSchemaIsValid) {
      let fieldIndex = -1;
      let isFound = false;
      const isIndexUndefined = isUndefined(index);

      while (fieldIndex++ < fields.length) {
        const isLast = fieldIndex === fields.length - 1;
        const isCurrentIndex =
          (isArray(index) ? index : [index]).indexOf(fieldIndex) >= 0;

        if (isCurrentIndex || isIndexUndefined) {
          isFound = true;
        }

        if (!isFound) {
          continue;
        }

        for (const key in fields[fieldIndex]) {
          const currentFieldName = `${name}[${fieldIndex}].${key}`;

          if (isCurrentIndex || isLast || isIndexUndefined) {
            validFieldsRef.current.delete(currentFieldName);
            fieldsWithValidationRef.current.delete(currentFieldName);
          } else {
            const previousFieldName = `${name}[${fieldIndex - 1}].${key}`;

            if (validFieldsRef.current.has(currentFieldName)) {
              validFieldsRef.current.add(previousFieldName);
            }
            if (fieldsWithValidationRef.current.has(currentFieldName)) {
              fieldsWithValidationRef.current.add(previousFieldName);
            }
          }
        }
      }

      shouldRender = true;
    }

    if (shouldRender) {
      reRender();
    }
  };

  const insert = (
    index: number,
    value: Partial<FormArrayValues> | Partial<FormArrayValues>[],
  ) => {
    mapCurrentFieldsValueWithState();
    resetFields(insertAt(getFieldValueByName(fieldsRef.current, name), index));
    commonTasks(
      insertAt(
        allFields.current,
        index,
        isArray(value) ? appendValueWithKey(value) : [appendId(value, keyName)],
      ),
    );

    if (errorsRef.current[name]) {
      errorsRef.current[name] = insertAt(
        errorsRef.current[name],
        index,
        fillEmptyArray(value),
      );
    }

    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      touchedFieldsRef.current[name] = insertAt(
        touchedFieldsRef.current[name],
        index,
        fillEmptyArray(value),
      );
      reRender();
    }
  };

  const swap = (indexA: number, indexB: number) => {
    mapCurrentFieldsValueWithState();
    const fieldValues = getFieldValueByName(fieldsRef.current, name);
    swapArrayAt(fieldValues, indexA, indexB);
    resetFields(fieldValues);
    swapArrayAt(allFields.current, indexA, indexB);
    commonTasks([...allFields.current]);

    if (errorsRef.current[name]) {
      swapArrayAt(errorsRef.current[name], indexA, indexB);
    }

    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      swapArrayAt(touchedFieldsRef.current[name], indexA, indexB);
      reRender();
    }
  };

  const move = (from: number, to: number) => {
    mapCurrentFieldsValueWithState();
    const fieldValues = getFieldValueByName(fieldsRef.current, name);
    moveArrayAt(fieldValues, from, to);
    resetFields(fieldValues);
    moveArrayAt(allFields.current, from, to);
    commonTasks([...allFields.current]);

    if (errorsRef.current[name]) {
      moveArrayAt(errorsRef.current[name], from, to);
    }

    if (readFormStateRef.current.touched && touchedFieldsRef.current[name]) {
      moveArrayAt(touchedFieldsRef.current[name], from, to);
      reRender();
    }
  };

  const reset = () => {
    resetFields();
    memoizedDefaultValues.current = get(defaultValuesRef.current, name, []);
    setField(mapIds(memoizedDefaultValues.current, keyName));
  };

  useEffect(() => {
    for (const watchField of watchFieldsRef.current) {
      if (watchField.startsWith(name)) {
        reRender();
        break;
      }
    }
  }, [fields, name, reRender, watchFieldsRef]);

  useEffect(() => {
    const resetFunctions = resetFieldArrayFunctionRef.current;
    const fieldArrayNames = fieldArrayNamesRef.current;
    fieldArrayNames.add(name);
    resetFunctions[name] = reset;

    return () => {
      resetFields();
      delete resetFunctions[name];
      fieldArrayNames.delete(name);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return {
    swap: useCallback(swap, []),
    move: useCallback(move, []),
    prepend: useCallback(prepend, []),
    append: useCallback(append, []),
    remove: useCallback(remove, [fields]),
    insert: useCallback(insert, []),
    fields,
  };
};
