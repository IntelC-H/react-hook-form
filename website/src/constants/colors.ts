export default {
  white: '#fff',
  lightBlue: '#3562cd',
  blue: '#244BA8',
};
