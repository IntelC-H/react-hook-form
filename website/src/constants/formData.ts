export default [
  {
    name: 'First name',
    type: 'text',
    required: true,
    max: '',
    min: '',
    maxLength: '80',
    minLength: '',
    pattern: '',
  },
  {
    name: 'Last name',
    type: 'text',
    required: true,
    max: '',
    min: '',
    maxLength: '100',
    minLength: '',
    pattern: '',
  },
  {
    name: 'Email',
    type: 'text',
    required: true,
    max: '',
    min: '',
    maxLength: '',
    minLength: '',
    pattern:
      '^(([^<>()\\[\\]\\\\.,;:\\s@"]+(\\.[^<>()\\[\\]\\\\.,;:\\s@"]+)*)|(".+"))@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$',
  },
  {
    name: 'Mobile number',
    type: 'tel',
    required: true,
    max: '11',
    min: '8',
    maxLength: '',
    minLength: '',
    pattern: '',
  },
  {
    name: 'Title',
    type: 'select',
    required: true,
    max: '',
    min: '',
    maxLength: '',
    minLength: '',
    pattern: '',
    options: 'Mr;Mrs;Miss;Dr',
  },
  {
    name: 'Developer',
    type: 'radio',
    required: true,
    max: '',
    min: '',
    maxLength: '',
    minLength: '',
    pattern: '',
    options: 'Yes;No',
  },
];
