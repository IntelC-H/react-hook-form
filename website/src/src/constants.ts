export const DATE_INPUTS = ['date', 'time', 'month', 'datetime', 'datetime-local', 'week'];

export const STRING_INPUTS = ['text', 'email', 'password', 'search', 'tel', 'url', 'textarea'];
