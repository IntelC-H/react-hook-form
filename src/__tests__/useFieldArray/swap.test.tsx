import * as React from 'react';
import {
  act as actComponent,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { act, renderHook } from '@testing-library/react-hooks';

import { VALIDATION_MODE } from '../../constants';
import { useFieldArray } from '../../useFieldArray';
import { useForm } from '../../useForm';
import { mockGenerateId } from '../useFieldArray.test';

describe('swap', () => {
  beforeEach(() => {
    mockGenerateId();
  });

  it('should swap into pointed position', () => {
    const { result } = renderHook(() => {
      const { register, control } = useForm({
        defaultValues: { test: [{ value: '1' }] },
      });
      const methods = useFieldArray({
        control,
        name: 'test',
      });

      return { register, ...methods };
    });

    act(() => {
      result.current.append({ value: '2' });
    });

    act(() => {
      result.current.swap(0, 1);
    });

    expect(result.current.fields).toEqual([
      { id: '1', value: '2' },
      { id: '0', value: '1' },
    ]);
  });

  it('should swap data order', () => {
    const { result } = renderHook(() => {
      const { register, control } = useForm({
        defaultValues: { test: [{ value: '1' }] },
      });
      const methods = useFieldArray({
        control,
        name: 'test',
      });

      return { register, ...methods };
    });

    act(() => {
      result.current.append({ value: '2' });
    });

    act(() => {
      result.current.swap(0, 1);
    });

    expect(result.current.fields).toEqual([
      { id: '1', value: '2' },
      { id: '0', value: '1' },
    ]);
  });

  it.each(['isDirty', 'dirtyFields'])(
    'should swap dirtyFields order when formState.%s is defined',
    () => {
      const { result } = renderHook(() => {
        const { formState, control } = useForm({
          defaultValues: { test: [{ value: '1' }] },
        });
        const methods = useFieldArray({
          control,
          name: 'test',
        });
        return {
          formState,
          ...methods,
        };
      });

      result.current.formState.isDirty;
      result.current.formState.dirtyFields;

      act(() => {
        result.current.append({ value: '2' });
      });

      act(() => {
        result.current.append({ value: '3' });
      });

      act(() => {
        result.current.swap(0, 1);
      });

      expect(result.current.formState.isDirty).toBeTruthy();
      expect(result.current.formState.dirtyFields).toEqual({
        test: [{ value: true }, { value: true }, { value: true }],
      });
    },
  );

  it('should swap errors', async () => {
    let errors: any;
    const Component = () => {
      const { register, handleSubmit, control, ...rest } = useForm({
        defaultValues: { test: [{ value: 'test' }] },
      });
      const { fields, append, swap } = useFieldArray({
        control,
        name: 'test',
      });
      errors = rest.formState.errors;

      return (
        <form onSubmit={handleSubmit(() => {})}>
          {fields.map((field, i) => (
            <input
              key={field.id}
              {...register(`test.${i}.value` as const, { required: true })}
            />
          ))}
          <button type="button" onClick={() => append({ value: '' })}>
            append
          </button>
          <button type="button" onClick={() => swap(0, 1)}>
            swap
          </button>
          <button>submit</button>
        </form>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button', { name: /append/i }));

    await actComponent(async () => {
      fireEvent.click(screen.getByRole('button', { name: /submit/i }));
    });

    expect(errors.test[0]).toBeUndefined();
    expect(errors.test[1]).toBeDefined();

    fireEvent.click(screen.getByRole('button', { name: /swap/i }));

    expect(errors.test[0]).toBeDefined();
    expect(errors.test[1]).toBeUndefined();
  });

  it('should swap touched fields', async () => {
    let touched: any;
    const Component = () => {
      const { register, formState, control } = useForm({
        defaultValues: { test: [{ value: 'test' }] },
      });
      const { fields, append, swap } = useFieldArray({
        control,
        name: 'test',
      });

      touched = formState.touchedFields;

      return (
        <form>
          {fields.map((field, i) => (
            <input key={field.id} {...register(`test.${i}.value` as const)} />
          ))}
          <button type="button" onClick={() => append({ value: '' })}>
            append
          </button>
          <button type="button" onClick={() => swap(0, 1)}>
            swap
          </button>
          <button>submit</button>
        </form>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button', { name: /append/i }));

    fireEvent.blur(screen.getAllByRole('textbox')[0]);

    fireEvent.click(screen.getByRole('button', { name: /swap/i }));

    expect(touched).toEqual({
      test: [undefined, { value: true }],
    });
  });

  it('should trigger reRender when user is watching the all field array', () => {
    const watched: any[] = [];
    const Component = () => {
      const { register, watch, control } = useForm({
        defaultValues: {
          test: [{ value: '1' }, { value: '2' }],
        },
      });
      const { fields, swap } = useFieldArray({
        control,
        name: 'test',
      });
      watched.push(watch());

      return (
        <form>
          {fields.map((field, i) => (
            <input key={field.id} {...register(`test.${i}.value` as const)} />
          ))}
          <button type="button" onClick={() => swap(0, 1)}>
            swap
          </button>
        </form>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button', { name: 'swap' }));

    expect(watched).toEqual([
      { test: [{ value: '1' }, { value: '2' }] }, // first render
      { test: [{ value: '2' }, { value: '1' }] }, // render inside useEffect in useFieldArray
      { test: [{ value: '2' }, { value: '1' }] }, // render inside swap method
      { test: [{ value: '2' }, { value: '1' }] }, // render inside useEffect in useFieldArray
    ]);
  });

  it('should return watched value with watch API', async () => {
    const renderedItems: any = [];
    const Component = () => {
      const { watch, register, control } = useForm<{
        test: {
          value: string;
        }[];
      }>();
      const { fields, append, swap } = useFieldArray({
        name: 'test',
        control,
      });
      const watched = watch('test');
      const isSwapped = React.useRef(false);
      if (isSwapped.current) {
        renderedItems.push(watched);
      }
      return (
        <div>
          {fields.map((field, i) => (
            <div key={`${field.id}`}>
              <input {...register(`test.${i}.value` as const)} />
            </div>
          ))}
          <button onClick={() => append({ value: '' })}>append</button>
          <button
            onClick={() => {
              swap(0, 1);
              isSwapped.current = true;
            }}
          >
            swap
          </button>
        </div>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button', { name: /append/i }));
    fireEvent.click(screen.getByRole('button', { name: /append/i }));

    const inputs = screen.getAllByRole('textbox');

    fireEvent.input(inputs[0], {
      target: { name: 'test[0].value', value: '111' },
    });
    fireEvent.input(inputs[1], {
      target: { name: 'test[1].value', value: '222' },
    });

    fireEvent.click(screen.getByRole('button', { name: /swap/i }));

    await waitFor(() =>
      expect(renderedItems).toEqual([
        [{ value: '222' }, { value: '111' }],
        [{ value: '222' }, { value: '111' }],
      ]),
    );
  });

  describe('with resolver', () => {
    it('should invoke resolver when formState.isValid true', async () => {
      const resolver = jest.fn().mockReturnValue({});

      const { result } = renderHook(() => {
        const { formState, control } = useForm({
          mode: VALIDATION_MODE.onChange,
          resolver,
          defaultValues: {
            test: [{ value: '1' }, { value: '2' }],
          },
        });
        const { swap } = useFieldArray({ control, name: 'test' });
        return { formState, swap };
      });

      result.current.formState.isValid;

      await act(async () => {
        result.current.swap(0, 1);
      });

      expect(resolver).toBeCalledWith(
        {
          test: [{ value: '2' }, { value: '1' }],
        },
        undefined,
        {
          criteriaMode: undefined,
          fields: {},
          names: [],
        },
      );
    });

    it('should not invoke resolver when formState.isValid false', () => {
      const resolver = jest.fn().mockReturnValue({});

      const { result } = renderHook(() => {
        const { formState, control } = useForm({
          mode: VALIDATION_MODE.onChange,
          resolver,
          defaultValues: {
            test: [{ value: '1' }, { value: '2' }],
          },
        });
        const { swap } = useFieldArray({ control, name: 'test' });
        return { formState, swap };
      });

      act(() => {
        result.current.swap(0, 1);
      });

      expect(resolver).not.toBeCalled();
    });
  });
});
