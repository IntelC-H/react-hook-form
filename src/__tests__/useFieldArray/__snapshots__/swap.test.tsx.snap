// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFieldArray array of array fields should populate all array fields correctly with setValue 1`] = `
<DocumentFragment>
  <div>
    <div>
      <input
        name="nest.0.value"
        value="1"
      />
      <div>
        <input
          name="nest.0.nestedArray.0.value"
          value="1"
        />
      </div>
    </div>
    <div>
      <input
        name="nest.1.value"
        value="2"
      />
      <div>
        <input
          name="nest.1.nestedArray.0.value"
          value="1"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`useFieldArray array of array fields should populate all array fields with setValue when name match Field Array 1`] = `
<DocumentFragment>
  <div>
    <div>
      <input
        name="nest.0.value"
        value="1"
      />
      <div>
        <input
          name="nest.0.nestedArray.0.value"
          value="1"
        />
      </div>
    </div>
    <div>
      <input
        name="nest.1.value"
        value="2"
      />
      <div>
        <input
          name="nest.1.nestedArray.0.value"
          value="1"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`useFieldArray array of array fields should worked with deep nested field array without chaining useFieldArray 1`] = `
<DocumentFragment>
  <div>
    <div>
      <input
        name="nest.0.value"
        value="1"
      />
      <div>
        <input
          name="nest.0.nestedArray.deepNest.0.value"
          value="1"
        />
        <button
          type="button"
        >
          append
        </button>
      </div>
      <button
        type="button"
      >
        setValue
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`useFieldArray array of array fields should worked with deep nested field array without chaining useFieldArray 2`] = `
<DocumentFragment>
  <div>
    <div>
      <input
        name="nest.0.value"
        value="1"
      />
      <div>
        <input
          name="nest.0.nestedArray.deepNest.0.value"
          value="1"
        />
        <input
          name="nest.0.nestedArray.deepNest.1.value"
          value="2"
        />
        <input
          name="nest.0.nestedArray.deepNest.2.value"
          value="3"
        />
        <button
          type="button"
        >
          append
        </button>
      </div>
      <button
        type="button"
      >
        setValue
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`useFieldArray array of array fields should worked with deep nested field array without chaining useFieldArray 3`] = `
<DocumentFragment>
  <div>
    <div>
      <input
        name="nest.0.value"
        value="1"
      />
      <div>
        <input
          name="nest.0.nestedArray.deepNest.0.value"
          value="1"
        />
        <input
          name="nest.0.nestedArray.deepNest.1.value"
          value="2"
        />
        <input
          name="nest.0.nestedArray.deepNest.2.value"
          value="3"
        />
        <input
          name="nest.0.nestedArray.deepNest.3.value"
          value="test"
        />
        <button
          type="button"
        >
          append
        </button>
      </div>
      <button
        type="button"
      >
        setValue
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`useFieldArray should append multiple inputs correctly 1`] = `
Array [
  Object {
    "test": Array [
      Object {
        "value": "data",
      },
    ],
  },
  Object {
    "test": Array [
      Object {
        "value": "data",
      },
    ],
  },
  Object {
    "test": Array [
      Object {
        "value": "data",
      },
      Object {
        "value": "test",
      },
      Object {
        "value": "test1",
      },
    ],
  },
  Object {
    "test": Array [
      Object {
        "value": "data",
      },
      Object {
        "value": "test",
      },
      Object {
        "value": "test1",
      },
    ],
  },
]
`;

exports[`useFieldArray should custom register append, prepend and insert inputs with values 1`] = `
Array [
  Array [],
  Array [
    Object {
      "test": "append",
      "test1": "append",
    },
  ],
  Array [
    Object {
      "test": "append",
      "test1": "append",
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "insert",
      "test1": "insert",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "insert",
      "test1": "insert",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "insert",
      "test1": "insert",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "insert",
      "test1": "insert",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "insert",
      "test1": "insert",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "insert",
      "test1": "insert",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
    Object {
      "test": "insert",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "insert",
      "test1": "insert",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
  ],
  Array [
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
    Object {
      "test": "insert",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
    Object {
      "test": "prepend",
      "test1": "prepend",
    },
    Object {
      "test": "insert",
      "test1": "insert",
    },
    Object {
      "test": "append",
      "test1": "append",
    },
    Object {
      "test": "prepend",
      "test2": Array [
        Object {
          "test": "test",
        },
      ],
    },
  ],
]
`;

exports[`useFieldArray should update field array defaultValues when invoke setValue 1`] = `
Array [
  Object {
    "names": Array [
      Object {
        "name": "will",
      },
      Object {
        "name": "Mike",
      },
    ],
  },
  Object {
    "names": Array [
      Object {
        "name": "will",
      },
      Object {
        "name": "Mike",
      },
    ],
  },
  Object {
    "names": Array [],
  },
  Object {
    "names": Array [],
  },
  Object {
    "names": Array [],
  },
]
`;
